content/courses/native-onchain-development/program-derived-addresses.md

```diff
@@ -22,258 +22,353 @@ description: "Get a deeper understanding of PDAs."
+amount of the program’s compute budget. Developers can optimize the performance
+and lower the compute budget of programs by passing the
+`bump_seed (canonical bump)`, the user-supplied seeds as part of the instruction
+data, and then deserialize the seed and canonical bump. These deserialized
+outputs can then be passed to the `create_program_address` method to derive the
+PDA. It's important to note that the `create_program_address` method incurs a
+fixed cost to the compute budget.
+
+Address collisions can occur since the seeds are passed as a slice of bytes,
+meaning that the seeds `{abcdef}`, `{abc, def}` and `{ab, cd, ef}` will result
+in the same PDA being generated. A developer must prevent collisions by adding
+separator characters like hyphens.
+
+In summary, the `find_program_address` method passes the input seeds and
+`program_id` to the `try_find_program_address` method. The
+`try_find_program_address` method starts with a `bump_seed` of 255, adds it to
+the input seeds, and then repeatedly calls the `create_program_address` method
+until it finds a valid PDA. Once found, both the PDA and the `bump_seed` are
+returned.
+
+Note that different valid bumps generate different valid PDAs for the same input
+seeds. The `bump_seed` returned by `find_program_address` will always be the
+first valid PDA found.
+
+Using the canonical bump when generating a PDA onchain is crucial. Starting with
+a `bump_seed` value of `255` and iterating downward to `0` ensures that the
+returned seed will always be the most significant valid `8-bit` value possible.
+This `bump_seed` is commonly known as the "_canonical bump_". It's a best
+practice always to use the canonical bump and validate every PDA passed into
+your program to ensure the integrity of the process.
+
+One point to emphasize is that the `find_program_address` method only returns a
+Program-Derived Address and the bump seed used to derive it. The method does not
+initialize a new account, nor is any PDA returned by the method necessarily
+associated with an account that stores data.
 
 ### Use PDA accounts to store data
 
-Since programs themselves are stateless, program state is managed through
-external accounts. Given that you can use seeds for mapping and that programs
-can sign on their behalf, using PDA accounts to store data related to the
-program is an extremely common design choice. While programs can invoke the
-System Program to create non-PDA accounts and use those to store data as well,
-PDAs tend to be the way to go.
+Solana programs are stateless, so they manage their state through external
+accounts. Although programs can use the System Program to create non-PDA
+accounts for data storage, PDAs are the choice for storing program-related data.
+This choice is popular because the seeds and canonical bump directly map to the
+same PDA, and the program specified as the program ID can sign on its behalf.
```

Comment:
```suggestion
Solana programs are stateless, so state is stored in separate accounts from where the program's executable is stored. Although programs can use the System Program to create non-PDA
accounts for data storage, PDAs are the choice for storing program-related data.
This choice is popular because the seeds and canonical bump directly map to the
same PDA, and the program specified as the program ID can sign on its behalf.
```

Reply:
Old language could be clearer.